tests/storage/clone.go

```diff
@@ -447,6 +451,133 @@ var _ = SIGDescribe("[Serial]VirtualMachineClone Tests", func() {
 				Expect(err).To(HaveOccurred())
 				Expect(err.Error()).Should(ContainSubstring("does not support snapshots"))
 			})
+
+			Context("with instancetype and preferences", func() {
+
+				var (
+					instancetype *instancetypev1alpha2.VirtualMachineInstancetype
+					preference   *instancetypev1alpha2.VirtualMachinePreference
+				)
+
+				checkForVMInstancetypeRevisionNames := func(vmNamespace, vmName string) func() error {
+					return func() error {
+						vm, err := virtClient.VirtualMachine(vmNamespace).Get(vmName, &metav1.GetOptions{})
+						if err != nil {
+							return err
+						}
+						if vm.Spec.Instancetype.RevisionName == "" {
+							return fmt.Errorf("instancetype revision name is expected to not be empty")
+						}
+
+						if vm.Spec.Preference.RevisionName == "" {
+							return fmt.Errorf("preference revision name is expected to not be empty")
+						}
+						return nil
+					}
+				}
+
+				waitForVMInstanceTypeRevisionNames := func(vmName string) {
+					Eventually(checkForVMInstancetypeRevisionNames(util.NamespaceTestDefault, vmName), 300*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+				}
+
+				BeforeEach(func() {
+					snapshotStorageClass, err := getSnapshotStorageClass(virtClient)
+					Expect(err).ToNot(HaveOccurred())
+
+					if snapshotStorageClass == "" {
+						Skip("Skipping test, no VolumeSnapshot support")
+					}
+
+					instancetype = &instancetypev1alpha2.VirtualMachineInstancetype{
+						ObjectMeta: metav1.ObjectMeta{
+							GenerateName: "vm-instancetype-",
+							Namespace:    util.NamespaceTestDefault,
+						},
+						Spec: instancetypev1alpha2.VirtualMachineInstancetypeSpec{
+							CPU: instancetypev1alpha2.CPUInstancetype{
+								Guest: 1,
+							},
+							Memory: instancetypev1alpha2.MemoryInstancetype{
+								Guest: resource.MustParse("128Mi"),
+							},
+						},
+					}
+					instancetype, err := virtClient.VirtualMachineInstancetype(util.NamespaceTestDefault).Create(context.Background(), instancetype, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					preference = &instancetypev1alpha2.VirtualMachinePreference{
+						ObjectMeta: metav1.ObjectMeta{
+							GenerateName: "vm-preference-",
+							Namespace:    util.NamespaceTestDefault,
+						},
+						Spec: instancetypev1alpha2.VirtualMachinePreferenceSpec{
+							CPU: &instancetypev1alpha2.CPUPreferences{
+								PreferredCPUTopology: instancetypev1alpha2.PreferSockets,
+							},
+						},
+					}
+					preference, err := virtClient.VirtualMachinePreference(util.NamespaceTestDefault).Create(context.Background(), preference, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					sourceVM = tests.NewRandomVMWithDataVolumeWithRegistryImport(
```

Comment:
Is it possible to use `libvmi` here?

Reply:
This is actually using libvmi under the hood there's just a boat load of complexity here because of the use of `DataVolumes`:

https://github.com/kubevirt/kubevirt/blob/73d78d58ca6b48c8d88ce3ec5a03e3e6355689bd/tests/utils.go#L502-L521